### ISSUE TITLE
xibo-player occasional crash while downloading files

### ISSUE THREAD

**Issue by dmitttri** [xibosignage/xibo-linux#256]
While performing requiested file download, xibo-player may crash with "Segmentation Fault" or "douible free or corruption"

It happens occasionaly, and is not easy repeateble. To reproduce it, I've made a procedure to manually delete cached contents, remove cacheFile.xml and schedule.xml, then start xibo-player without watchdog, just to print logs to stdout. This way xibo-player is forced to request all cache files again form CMS. Sometimes we can observe segfaults, but most of the time download process goes fine.

Log sample:
[08:58:42.149] [28723] [debug]: [XMDS::RequiredFiles] Received
[08:58:47.513] [28762] [debug]: [jquery-1.11.1.min.js] Downloaded
[08:58:47.520] [28764] [debug]: [18.otf] Downloaded
[08:58:47.520] [28763] [debug]: [368.html] Downloaded
[08:58:47.521] [28765] [debug]: [369.html] Downloaded
[08:58:52.849] [28791] [debug]: [16.ttf] Downloaded
[08:58:52.878] [28792] [debug]: [14.otf] Downloaded
[08:58:52.885] [28793] [debug]: [xibo-countdown-render.js] Downloaded
[08:58:52.920] [28794] [debug]: [jquery-cycle-2.1.6.min.js] Downloaded
Segmentation fault

I seems we can trigger segfaults more often  with CMS layouts which has lot of small files. 

- Linux Disto: Custom Buildroot Linux
- Linux Desktop Environment: Weston
- How the player was launched: compiled binaries

**Comment by dmitttri** [comment]
Further investigation shows that downloader threads will concurrently access FileCache_ object, more particular it's save() method, which saves files and updates cache. In particular, updating file cache was causing this issue.

I made a fix for this problem, by adding a mutex in the FileCacheImpl class, and eure addToCache method use it to protect file cache access. For this fix I have submitted a pull request.
